### PR TITLE
[SPARK-34087][SQL] Fix memory leak of ExecutionListenerBus

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR proposes an alternative way to fix the memory leak of `ExecutionListenerBus`, which would automatically clean them up.

Basically, the idea is to add `registerSparkListenerForCleanup` to `ContextCleaner`, so we can remove the `ExecutionListenerBus` from `LiveListenerBus` when the `SparkSession` is GC'ed.

On the other hand, to make the `SparkSession` GC-able, we need to get rid of the reference of `SparkSession` in `ExecutionListenerBus`. Therefore, we introduced the `sessionUUID`, which is a unique identifier for SparkSession, to replace the  `SparkSession` object.

Note that, the proposal wouldn't take effect when `spark.cleaner.referenceTracking=false` since it depends on `ContextCleaner`.

### Why are the changes needed?

Fix the memory leak caused by `ExecutionListenerBus` mentioned in SPARK-34087.

### Does this PR introduce _any_ user-facing change?

Yes, save memory for users.

### How was this patch tested?

Added unit test.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan

File: core/src/main/scala/org/apache/spark/ContextCleaner.scala [hunk 1]
```diff
@@ -293,4 +314,5 @@ private[spark] trait CleanerListener {
   def broadcastCleaned(broadcastId: Long): Unit
   def accumCleaned(accId: Long): Unit
   def checkpointCleaned(rddId: Long): Unit
+  def listenerCleaned(listener: SparkListener): Unit
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/ContextCleaner.scala:317]: looking at other methods in this trait, they all take IDs. Should this method take session id?

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -102,6 +103,8 @@ class SparkSession private(
         new SparkSessionExtensions), Map.empty)
   }
 
+  val sessionUUID: String = UUID.randomUUID.toString
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:106]: shall we mark it as `private[sql]` if we don't want to expose it as a public API?

**Comment by cloud-fan** [comment]
The overall approach LGTM

File: core/src/main/scala/org/apache/spark/ContextCleaner.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/ContextCleaner.scala:317]: The session id doesn't have a 1:1 relationship with the listener so I don't think we should use the session id.

File: core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala [hunk 3]
```diff
@@ -31,6 +31,7 @@ import org.scalatest.time.SpanSugar._
 import org.apache.spark.internal.{config, Logging}
 import org.apache.spark.internal.config._
 import org.apache.spark.rdd.{RDD, ReliableRDDCheckpointData}
+import org.apache.spark.scheduler.SparkListener
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala:34]: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 4]
```diff
@@ -19,18 +19,22 @@ package org.apache.spark.sql
...
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.internal.StaticSQLConf._
+import org.apache.spark.sql.util.ExecutionListenerBus
 import org.apache.spark.util.ThreadUtils
 
 /**
  * Test cases for the builder pattern of [[SparkSession]].
  */
-class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
+class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach with Eventually {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:37]: unnecessary changes

**Comment by gengliangwang** [comment]
Nice fix! How about adding a test case for this? (the test in "manual test" seems fine)

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala [hunk 5]
```diff
@@ -76,7 +76,11 @@ trait QueryExecutionListener {
 class ExecutionListenerManager private[sql](session: SparkSession, loadExtensions: Boolean)
   extends Logging {
 
-  private val listenerBus = new ExecutionListenerBus(session)
+  private val listenerBus = new ExecutionListenerBus(session.sessionUUID)
+  session.sparkContext.listenerBus.addToSharedQueue(listenerBus)
+  session.sparkContext.cleaner.foreach { cleaner =>
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala:81]: So, this fix is designed to be disabled when `spark.cleaner.referenceTracking=false`?

File: core/src/main/scala/org/apache/spark/ContextCleaner.scala [hunk 6]
```diff
@@ -276,6 +287,15 @@ private[spark] class ContextCleaner(
     }
   }
 
+  def doCleanSparkListener(listener: SparkListener): Unit = {
+    try {
+      logDebug(s"Cleaning Spark listener $listener")
+      sc.listenerBus.removeListener(listener)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/ContextCleaner.scala:293]: Maybe, shall we add a post-log?
```scala
logDebug(s"Cleaned Spark listener $listener")
```

**Comment by Ngone51** [comment]
> Nice fix! How about adding a test case for this? (the test in "manual test" seems fine)

Added it as the unit test. @gengliangwang

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 5 above)
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala:81]: To be honest, no. It fixes the memory leak so we generally always want it takes effect. And the `ContextCleaner` is an experienced framework that we can leverage to fix this issue elegantly without aggressive changes.

The config is introduced since Spark 1.0.0 and enabled by default. So I personally think it's acceptable to assume users run applications with cleaner enabled in most cases.

File: core/src/main/scala/org/apache/spark/ContextCleaner.scala [hunk 7]
```diff
@@ -276,6 +287,16 @@ private[spark] class ContextCleaner(
     }
   }
 
+  def doCleanSparkListener(listener: SparkListener): Unit = {
+    try {
+      logDebug(s"Cleaning Spark listener $listener")
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/ContextCleaner.scala:292]: Does logging `$listener` show meaningful message? It is not like other cases like clean up RDD (rdd id), or clean up broadcast (broadcast id).

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/ContextCleaner.scala:292]: It logs the qualified class name and JVM object id of that listener, e.g.,

> Cleaning Spark listener org.apache.spark.sql.util.ExecutionListenerBus@54600fe

I think it's enough for users to know which listener is removed.

Others log their own ids because users can track them by analyzing the logs since Spark logs those ids in other places too. But Spark doesn't do that for listeners (and the listener itself doesn't have such an identifier or something else) so it should be enough for now.

**Review by viirya — COMMENTED** [review]
The approach looks making sense.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 8]
```diff
@@ -40,6 +45,37 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
...
+    }
+
+    (1 to 10).foreach { _ =>
+      spark.cloneSession()
+      SparkSession.clearActiveSession()
+    }
+
+    var num = listenersNum()
+    // Before GC, the number of ExecutionListenerBus is (1 + 10)
+    assert(num === 11)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:68]: This is unreliable, the GC may happen at any time, even before this assert.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 9]
```diff
@@ -40,6 +45,37 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
...
+    eventually(timeout(10.seconds), interval(1.seconds)) {
+      System.gc()
+      num = listenersNum()
+      // After GC, the number of ExecutionListenerBus should be less than 11
+      // Since GC can't 100% guarantee all out-of-referenced objects be cleaned at one time,
+      // here, we check at least one listener is cleaned up to prove the mechanism works.
+      assert(num < 11)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:75]: shall we simply check `assert(num < 10)`? We created 10 sessions and if `num < 10`, it proves the listeners from the session are GCed.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 10]
```diff
@@ -40,6 +45,37 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
     SparkSession.clearDefaultSession()
   }
 
+  test("SPARK-34087: Fix memory leak of ExecutionListenerBus") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:48]: how long does this test take?

Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:48]: 3s+

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala (hunk 9 above)
Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:75]: We actually create 11 sessions: original one and 10 cloned.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala (hunk 8 above)
Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:68]: make sense

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala:81]: I'm not against having that assumption. But, this doesn't fix memory leaks at all if `spark.cleaner.referenceTracking=false` technically, does it? We should be clear about the limitation of the PR.

Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala:81]: > But, this doesn't fix memory leaks at all if spark.cleaner.referenceTracking=false technically, does it? 

Yes, it doesn't.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala:81]: Let's mention it in the PR description to be clear.

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala (hunk 8 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:68]: let's remove this line then.

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 5 above)
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala:81]: Updated.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 11]
```diff
@@ -40,6 +45,36 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
...
+        .asScala
+        .count(_.isInstanceOf[ExecutionListenerBus])
+    }
+
+    (1 to 10).foreach { _ =>
+      spark.cloneSession()
+      SparkSession.clearActiveSession()
+    }
+
+    var num = listenersNum()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:66]: we can move it inside `eventually {...}` and inline it
```
eventually ... {
  ...
  assert(listenersNum() < 11)
}
```

**Review by cloud-fan — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]
LGTM

**Comment by gengliangwang** [comment]
@Ngone51 I have noticed you are already a member of the Apache Spark project. Congratulations!
Would you try merging this one by yourself?

**Comment by Ngone51** [comment]
Sure, I'd like to have a try after tests pass.

**Review by viirya — APPROVED** [review]

**Review by cfmcgrady — APPROVED** [review]

**Comment by Ngone51** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

@Ngone51 it has a conflict with branch-3.1. Would you mind creating a PR to backport?

**Comment by Ngone51** [comment]
Sure.
